added rng_guide (#3497)
google/flax: docs/guides/flax_fundamentals/rng_guide.md

---
jupytext:
  formats: ipynb,md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.13.8
---

# Working with RNG's

+++

This guide will demonstrate how Flax uses Jax's PRNG keys to emulate randomness and adds some additional features to make it easier for users to thread PRNG keys through different modules. For a quick refresher on Jax PRNG keys, check out [Pseudo Random Numbers in JAX](https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html) and [🔪 JAX - The Sharp Bits 🔪](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#random-numbers).

+++

## Setup

+++

Import some necessary dependencies.

**Note:** This guide uses the `--xla_force_host_platform_device_count=8` flag to emulate multiple devices in a CPU environment in a Google Colab/Jupyter Notebook. You don’t need this if you are already using a multi-device TPU environment.

```{code-cell}

> !pip install -q flax

Reviewer: we need to add the `skip-execution` tag to this cell

    # same value as second code snippet above
    print(f"rng_stream2: {self.make_rng('rng_stream2')}")
    # same value as first code snippet above
    print(f"rng_stream1: {self.make_rng('rng_stream1')}")
    # same value as second code snippet above
    print(f"rng_stream2: {self.make_rng('rng_stream2')}")
    # same value as first code snippet above
    print(f"rng_stream1: {self.make_rng('rng_stream1')}")
    # same value as second code snippet above
    print(f"rng_stream2: {self.make_rng('rng_stream2')}")

rng_module_two_streams = RNGModuleTwoStreams()
variables = rng_module_two_streams.init({'rng_stream1': jax.random.key(0), 'rng_stream2': jax.random.key(1)})
```

Providing the same seed PRNG key will result in the same values being generated (provided that the same operations are used for those keys).

```{code-cell}
:outputId: 1f711ced-a97a-49e5-e92c-15ee6d57f93a

variables = rng_module_two_streams.init({'rng_stream1': jax.random.key(0), 'rng_stream2': jax.random.key(0)})
```

### How `self.make_rng` works under the hood

+++

When `self.make_rng` is called, the following happens:
* The following data is collected:

>   * the name of the module making the `self.make_rng` call (note: the top-level root module has no name)
>   * the name of all of its parent modules, all the way up until the top-level root module

Reviewer: ```suggestion
  * the path of the Module as given by `self.scope.path` (the top-level root module has an empty path `()`)
```

            Node(name="abc", leaf_name="def")(x),
    )

out1, out2 = Model().apply({}, 0, rngs={"rng": jax.random.PRNGKey(33)})
out1 == out2 # same output, despite having different submodule names
```

This occurs because the hash function [concatenates strings together](https://docs.python.org/3/library/hashlib.html#hashlib.hash.update), so the data `('AB', 'C')` is equivalent to data `('A', 'BC')` when fed into the hash function, therefore producing the same hash int.

```{code-cell}
:outputId: 1ed17f72-43cc-4f00-95af-5231f13a7f4b

print(produce_hash(data=('A', 'B', 'C', 1)))
print(produce_hash(data=('AB', 'C', 1)))
print(produce_hash(data=('A', 'BC', 1)))
print(produce_hash(data=('ABC', 1)))
```

To avoid this edge case, users can flip the `flax_fix_rng_separator` [configuration flag](https://flax.readthedocs.io/en/latest/api_reference/flax.config.html#flax.configurations.Config.flax_fix_rng_separator) to `True`.

```{code-cell}
:outputId: 21e623da-91d2-4bbc-e167-d3b257344124

flax.config.update('flax_fix_rng_separator', True)
out1, out2 = Model().apply({}, 0, rngs={"rng": jax.random.PRNGKey(33)})
out1 == out2 # different output
```

> ## Managing RNG's on multiple devices

Reviewer: Wondering if we should recommend `shard_map` instead?

Author: what about having both an example for `jit` and and for `shard_map`?

Reviewer 2: +1 to both `(p)jit` and `shard_map`

```

If we choose not to add noise, then the output is the same across all batches (as expected, since the input is the same for all batches):

```{code-cell}
:outputId: 235606f5-2d7d-404c-b2ce-fd60279f7605

out = jit_forward(variables, x, False, {'params': apply_rng})
print(out)
```

## Lifted transforms

+++

[Flax lifted transforms](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/transformations.html) allow you to use [jax transforms](https://github.com/google/jax#transformations) with `Module` arguments. This section will show you how to control how PRNG keys are split in Flax lifted transforms.

See [Lifted transformations](https://flax.readthedocs.io/en/latest/developer_notes/lift.html) for more detail.

+++

### `nn.vmap`

+++

We can use [`nn.vmap`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/_autosummary/flax.linen.vmap.html) to create a batched `Dense` layer:

```{code-cell}
:outputId: ba010fb0-e6ff-4e16-bdd7-203e4f88b155

x = jnp.ones((3, 2))

BatchDense = nn.vmap(
    nn.Dense,
    in_axes=0, out_axes=0,
    variable_axes={'params': None},
    split_rngs={'params': False})

BatchDense(2).init(jax.random.key(0), x)
```

> By denoting `variable_axes={'params': 0}'`, we can generate parameters for each batch input. However the parameter values generated are all identical to each other:

Reviewer: ```suggestion
By denoting `variable_axes={'params': 0}'`, vectorize the `params` Arrays on the first axis. However the parameter values generated are all identical to each other:
```

  @nn.compact
  def __call__(self):
    # same value as first code snippet above
    print(f"rng_stream1: {self.make_rng('rng_stream1')}")
    # same value as second code snippet above
    print(f"rng_stream2: {self.make_rng('rng_stream2')}")
    # same value as first code snippet above
    print(f"rng_stream1: {self.make_rng('rng_stream1')}")
    # same value as second code snippet above
    print(f"rng_stream2: {self.make_rng('rng_stream2')}")
    # same value as first code snippet above
    print(f"rng_stream1: {self.make_rng('rng_stream1')}")
    # same value as second code snippet above
    print(f"rng_stream2: {self.make_rng('rng_stream2')}")

rng_module_two_streams = RNGModuleTwoStreams()
variables = rng_module_two_streams.init({'rng_stream1': jax.random.key(0), 'rng_stream2': jax.random.key(1)})
```

Providing the same seed PRNG key will result in the same values being generated (provided that the same operations are used for those keys).

```{code-cell}
:outputId: 1f711ced-a97a-49e5-e92c-15ee6d57f93a

variables = rng_module_two_streams.init({'rng_stream1': jax.random.key(0), 'rng_stream2': jax.random.key(0)})
```

### How `self.make_rng` works under the hood

+++

> When `self.make_rng` is called, the following happens:

Reviewer 2: Suggestion:

```suggestion
When `self.make_rng` (`flax.linen.Module.make_rng`) is called, the following happens:
```

  * the `self.make_rng` call count; i.e. the number of times `self.make_rng` has been called for this specific stream (including this call)
    * note: each submodule will have its own individual call count that's separate from other modules; e.g. a module that has called `self.make_rng('params')` twice and contains a submodule that has called `self.make_rng('params')` once, will have a call count of 2 and 1 for the RNG stream `'params'`, respectively
* the data is bundled into a tuple and fed into a hash function and produces an int
* the generated int is folded into the RNG stream's starting seed PRNG key to generate a new, unique PRNG key

+++

This is a slightly simplified version of the hash function that Flax uses for `self.make_rng`:

```{code-cell}
def produce_hash(data):
  m = hashlib.sha1()
  for x in data:
    if isinstance(x, str):
      m.update(x.encode('utf-8'))
    elif isinstance(x, int):
      m.update(x.to_bytes((x.bit_length() + 7) // 8, byteorder='big'))
    else:
      raise ValueError(f'Expected int or string, got: {x}')
  d = m.digest()
  hash_int = int.from_bytes(d[:4], byteorder='big')
  return hash_int
```

> Following the instructions above, we can manually reproduce the PRNG keys generated from `self.make_rng`.

Reviewer 2: ```suggestion
Following the instructions above, you can manually reproduce the PRNG keys generated from `self.make_rng` as follows:
```

```{code-cell}
:outputId: 07048549-e011-4c3e-e614-9401b31d7dc9

stream_seed = jax.random.key(0)
for call_count in range(1, 4):
  hash_int = produce_hash(data=(call_count,))
  print(jax.random.fold_in(stream_seed, jnp.uint32(hash_int)))
```

```{code-cell}
:outputId: 89d7cce4-da43-4393-e07c-73f2c53cefe4

variables = rng_module.init({'rng_stream': jax.random.key(0)})
```

> ### Submodules

Reviewer 2: ```suggestion
### Sub-`Module`s
```

+++

> Let's take a look at how `self.make_rng` behaves with submodules.

Reviewer 2: ```suggestion
This section explores how `self.make_rng` (`flax.linen.Module.make_rng`) behaves with sub-`Module`s.

For example:
```

  * for `RNGModule`, the data is just the call count `(1,)` and `(2,)`, since the root module has an empty path
  * for `RNGSubModule`, the data is `('RNGSubModule_0', 1)` and `('RNGSubModule_0', 2)`
  * for `RNGSubSubModule`, the data is `('RNGSubModule_0', 'RNGSubSubModule_0', 1)` and `('RNGSubModule_0', 'RNGSubSubModule_0', 2)`

With this data, we are able to manually reproduce the PRNG keys generated from the module and submodules using `self.make_rng`:

```{code-cell}
:outputId: a8369212-d74b-491b-8aa9-d4782f98e6e8

stream_seed = jax.random.key(0)
for initial_data in ((), ('RNGSubModule_0',), ('RNGSubModule_0', 'RNGSubSubModule_0')):
  if initial_data:
    module_name = initial_data[-1]
  else:
    module_name = 'RNGModule'
  for call_count in (1, 2):
    hash_int = produce_hash(data=initial_data+(call_count,))
    rng_key = jax.random.fold_in(stream_seed, jnp.uint32(hash_int))
    print(f"{module_name}, count {call_count}: {rng_key}")
```

> If the same submodule class is used multiple times, we simply increment the suffix of the submodule name accordingly; i.e. `RNGSubModule_0`, `RNGSubModule_1`, etc.

Reviewer 2: Nit:

```suggestion
If the same sub-`Module` class is used multiple times, you can increment the suffix of the sub-`Module` name accordingly. For example: `RNGSubModule_0`, `RNGSubModule_1`, and so on.
```

```{code-cell}
:outputId: 0ab17746-6d21-4f46-8677-b1b3a6919231

class RNGSubModule(nn.Module):
  @nn.compact
  def __call__(self):
    print(f"{self.name}, count 1: {self.make_rng('rng_stream')}")
    print(f"{self.name}, count 2: {self.make_rng('rng_stream')}")

class RNGModule(nn.Module):
  @nn.compact
  def __call__(self):
    print(f"RNGModule, count 1: {self.make_rng('rng_stream')}")
    print(f"RNGModule, count 2: {self.make_rng('rng_stream')}")
    RNGSubModule()()
    RNGSubModule()()

rng_module = RNGModule()
variables = rng_module.init({'rng_stream': jax.random.key(0)})
```

```{code-cell}
:outputId: ff8a7eb6-80ee-4ba9-cf79-dff002a0aa43

stream_seed = jax.random.key(0)
for initial_data in ((), ('RNGSubModule_0',), ('RNGSubModule_1',)):
  if initial_data:
    module_name = initial_data[-1]
  else:
    module_name = 'RNGModule'
  for call_count in (1, 2):
    hash_int = produce_hash(data=initial_data+(call_count,))
    rng_key = jax.random.fold_in(stream_seed, jnp.uint32(hash_int))
    print(f"{module_name}, count {call_count}: {rng_key}")
```

### Using `self.param` and `self.variable`

+++

> Flax users have the option of creating additional parameters and variables in their modules by using the [`self.param`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/module.html#flax.linen.Module.param) and [`self.variable`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/module.html#flax.linen.Module.variable) `Module` methods. An `init_fn` argument must be passed to these methods so that it can generate the initial value of the parameter / variable. `self.make_rng` is commonly used implicitly or explicitly in this `init_fn`, since many initializer functions are stochastic in nature and require a PRNG key. See the full list of Flax initializers [here](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/initializers.html).

Reviewer 2: Nit:

```suggestion
You can create additional parameters and variables in their `Module`s by using the [`self.param`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/module.html#flax.linen.Module.param) and [`self.variable`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/module.html#flax.linen.Module.variable) `Module` methods.
- Remember: You must pass the `init_fn` argument these methods, so that it can generate the initial value of the parameter/variable.
- `self.make_rng` is commonly used implicitly or explicitly in the `init_fn` argument, since many initializer functions are stochastic in nature and require a PRNG key.

For a full list of Flax initializers, check out [`flax.linen.initializers`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/initializers.html).
```

print(variables['other_collection']['model_variable2'])
print(variables['params']['Dense_1']['kernel'])
print(variables['params']['Dense_1']['bias'])
```

Remember:
* there is a separate count for each RNG stream; this is why the count for `self.make_rng('other')` starts at 1 even though there were earlier calls of `self.make_rng('params')`
* each submodule has their own separate count for each rng stream; this is why each `Dense` layer has their own separate count for `self.make_rng('params')` and why `model_param` and `model_variable1` share the same count (since they are defined within the same top-level parent module)

```{code-cell}
:outputId: 44da3084-f9e8-478e-eb45-adb5cc8aafcc

params_seed = jax.random.key(0)
other_seed = jax.random.key(1)
for initial_data, count, seed, shape in (
  (('Dense_0',), 1, params_seed, (2, 2)),
  ((), 1, params_seed, (2, 2)),
  ((), 2, params_seed, (2, 2)),
  ((), 1, other_seed, (2, 2)),
  (('Dense_1',), 1, params_seed, (2, 2)),
  (('Dense_1',), 2, params_seed, (1, 2)),
):
  hash_int = produce_hash(data=(*initial_data, count))
  rng_key = jax.random.fold_in(seed, jnp.uint32(hash_int))
  print(jax.random.normal(rng_key, shape))
```

> ### Managing RNG streams in a training loop

Reviewer 2: ```suggestion
### Managing PRNG streams inside a training loop
```

      x,
      train=True,
      rngs={'dropout': dropout_rng},
    )
    return jnp.mean((y - out) ** 2)
  grads = jax.grad(loss)(variables['params'])
  params = jax.tree_map(lambda p, g: p - 1e-3 * g, variables['params'], grads)
  return {
    'params': params,
    'other_collection': variables['other_collection'],
  }, next_rng

for _ in range(10):
  variables, train_rng = update(variables, train_rng)
  out = module.apply(variables, x, train=False)
  print(jnp.mean((y - out)**2))
```

### 🔪 Sharp edge 🔪 - unintentionally generating the same values

There is an edge case where the same value can be unintentionally generated.
See the [Flax issue](https://github.com/google/flax/issues/2157) for more details.

```{code-cell}
:outputId: e4915d4b-591f-48d9-e03e-29ace2dfc4fe

class Leaf(nn.Module):
  def __call__(self, x):
    return x + jax.random.randint(self.make_rng("rng"), (), 0, 100)

class Node(nn.Module):
  leaf_name: str
  @nn.compact
  def __call__(self, x):
    return Leaf(name=self.leaf_name)(x)

class Model(nn.Module):
  @nn.compact
  def __call__(self, x):
    return (Node(name="ab", leaf_name="cdef")(x),
            Node(name="abc", leaf_name="def")(x),
    )

> out1, out2 = Model().apply({}, 0, rngs={"rng": jax.random.PRNGKey(33)})

Reviewer 3: ```suggestion
out1, out2 = Model().apply({}, 0, rngs={"rng": jax.random.key(33)})
```
Would this work?

out1 == out2 # same output, despite having different submodule names
```

This occurs because the hash function [concatenates strings together](https://docs.python.org/3/library/hashlib.html#hashlib.hash.update), so the data `('AB', 'C')` is equivalent to data `('A', 'BC')` when fed into the hash function, therefore producing the same hash int.

```{code-cell}
:outputId: 1b30c64d-4cf5-4792-f4ed-88b7e5cdd6b5

print(produce_hash(data=('A', 'B', 'C', 1)))
print(produce_hash(data=('AB', 'C', 1)))
print(produce_hash(data=('A', 'BC', 1)))
print(produce_hash(data=('ABC', 1)))
```

To avoid this edge case, users can flip the `flax_fix_rng_separator` [configuration flag](https://flax.readthedocs.io/en/latest/api_reference/flax.config.html#flax.configurations.Config.flax_fix_rng_separator) to `True`.

```{code-cell}
:outputId: d7905467-1b77-4f7f-f64c-607d4c97e1c8

flax.config.update('flax_fix_rng_separator', True)

> out1, out2 = Model().apply({}, 0, rngs={"rng": jax.random.PRNGKey(33)})

Reviewer 3: ```suggestion
out1, out2 = Model().apply({}, 0, rngs={"rng": jax.random.key(33)})
```
Here too.

- [JAX 101: PRNGs in JAX](https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html)
- [🔪 JAX - The Sharp Bits 🔪: Random Numbers](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#random-numbers).

+++

## Setup

+++

Install or upgrade Flax, and then import some necessary dependencies.

**Note:** This guide uses the `--xla_force_host_platform_device_count=8` flag to emulate multiple devices in a CPU environment in a Google Colab/Jupyter Notebook. You don’t need this if you are already using a multi-device Google Cloud TPU environment, for example, on Google Cloud or in a Kaggle VM with a TPU.

```{code-cell}
:outputId: bb13d0ba-f908-4746-e4d3-f5916e0670ff
:tags: [skip-execution]

!pip install -q flax
```

```{code-cell}
import os
os.environ["XLA_FLAGS"] = '--xla_force_host_platform_device_count=8'
```

```{code-cell}
import flax, flax.linen as nn
import jax, jax.numpy as jnp
from jax.sharding import Mesh, PartitionSpec, NamedSharding
from jax.experimental import mesh_utils
from jax.experimental.shard_map import shard_map

import hashlib
```

```{code-cell}
:outputId: ab5e0d3b-2d51-44ee-b823-d152ad1a10b2

jax.devices()
```

> Set Jax config variable `jax_threefry_partitionable` to `True`. This will be the default value in the future and makes the RNG more efficiently auto-parallelizable under `jax.jit`. See this [JAX discussion](https://github.com/google/jax/discussions/18480) for more details.

Reviewer 2: Nit:
```suggestion
Set the JAX config variable `jax_threefry_partitionable` to `True`. This will be the default value in the future and makes the PRNG more efficiently auto-parallelizable under `jax.jit`. Refre to this [JAX discussion](https://github.com/google/jax/discussions/18480) for more details.
```

```{code-cell}
:outputId: 29b2911f-28a7-4a25-fcf7-4c3544dafada

def forward(variables, x, add_noise, rng_key_batch):
  # rng_key_batch is a batch of size 1 containing 1 PRNG key
  # index slice into the rng_key_batch to access the PRNG key
  return module.apply(
    variables, x, add_noise, rngs={'params': rng_key_batch[0]}
  )

# define partition specifications
data_pspec = PartitionSpec('data')
no_pspec = PartitionSpec()

# shard the inputs x and rng keys across devices
# replicate the variables and add_noise boolean across devices
# shard the output across devices
shmap_forward = shard_map(
  forward,
  mesh=mesh,
  in_specs=(no_pspec, data_pspec, no_pspec, data_pspec),
  out_specs=data_pspec,
)
# get 8 different rng's that will be used by the 8 devices when doing forward inference
apply_rngs = jax.random.split(apply_rng, 8)
out = shmap_forward(variables, x, True, apply_rngs)
out
```

Confirm that the output is sharded across devices:

```{code-cell}
:outputId: 2d74dff9-3a7d-47df-afbc-3bda4b8edf51

out.addressable_shards
```

```{code-cell}
:outputId: 28bfe59b-ded1-455a-c894-f1c457ec28bf

jax.debug.visualize_array_sharding(out)
```

## Lifted transforms

+++

> [Flax lifted transforms](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/transformations.html) allow you to use [jax transforms](https://github.com/google/jax#transformations) with `Module` arguments. This section will show you how to control how PRNG keys are split in Flax lifted transforms.

Reviewer 2: ```suggestion
[Flax lifted transforms](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/transformations.html) allow you to use [JAX transforms](https://github.com/google/jax#transformations) with `Module` arguments. This section will show you how to control how PRNG keys are split in the Flax lifted transforms.
```
